pkg/storage/chunk/schema_config.go

```diff
@@ -446,3 +497,42 @@ func (cfg *PeriodicTableConfig) TableFor(t model.Time) string {
 func (cfg *PeriodicTableConfig) tableForPeriod(i int64) string {
 	return cfg.Prefix + strconv.Itoa(int(i))
 }
+
+// Generate the appropriate external key based on cfg.Schema, chunk.Checksum, and chunk.From
+func (cfg SchemaConfig) ExternalKey(chunk Chunk) string {
+	p, err := cfg.SchemaForTime(chunk.From)
+	v, _ := p.VersionAsInt()
+	if err == nil && v >= 12 {
+		return cfg.newerExternalKey(chunk, p.ChunkPathShardFactor, p.ChunkPathPeriod)
+	} else if chunk.ChecksumSet {
+		return cfg.newExternalKey(chunk)
+	} else {
+		return cfg.legacyExternalKey(chunk)
+	}
+}
+
+// pre-checksum
+func (cfg SchemaConfig) legacyExternalKey(chunk Chunk) string {
```

Comment:
It looks like there might be a bug in Loki currently where we prepend `UserID` to legacy keys but never successfully parse them. I discovered this with the introduction of new tests to cover all (3) variations of external key.

I don't quite understand the context here via the comment so if there is something I'm missing by removing this please chime in.

Existing [legacy key creation](https://github.com/grafana/loki/blob/main/pkg/storage/chunk/chunk.go#L214) vs. [legacy key parsing](https://github.com/grafana/loki/blob/main/pkg/storage/chunk/chunk.go#L104).

Reply:
Likely possible, you can send a separate PR if needed.